In [ ]:
# Get data from Google Drive
!gdown 1V7EiaJmwXvqdkVX_RSHMtgyPayTztp3a
!gdown 1Yc52N7WmdAqrFlh0d-aL2Gnex3G7KVUH

Downloading...
From: https://drive.google.com/uc?id=1V7EiaJmwXvqdkVX_RSHMtgyPayTztp3a
To: /content/202508_airbnb_listings_nyc.csv
100% 5.74M/5.74M [00:00<00:00, 87.7MB/s]
Downloading...
From: https://drive.google.com/uc?id=1Yc52N7WmdAqrFlh0d-aL2Gnex3G7KVUH
To: /content/202508_airbnb_listings_jc.csv
100% 346k/346k [00:00<00:00, 69.3MB/s]


In [ ]:
import pandas as pd
nyc_df = pd.read_csv('202508_airbnb_listings_nyc.csv')
jc_df = pd.read_csv('202508_airbnb_listings_jc.csv')

## Part 1: Building the High Risk Capital Partners (HRCP) Model


### Step 1:- Skiping Exploratory Data Analysis

In [ ]:
nyc_df.shape

(36403, 18)

In [ ]:
jc_df.shape

(1798, 18)

In [ ]:
y = nyc_df.availability_365
X = nyc_df.select_dtypes('number').drop(columns=['availability_365'])


In [ ]:
y.head()

0    365
1    365
2      1
3     24
4    288
Name: availability_365, dtype: int64

In [ ]:
nyc_df.head()

id                                               name  host_id  \
0  2539   Superfast Wi-Fi.  Clean & quiet home by the park     2787   
1  2595            Skylit Studio Oasis | Midtown Manhattan     2845   
2  6848                   Only 2 stops to Manhattan studio    15991   
3  6872  Uptown Sanctuary w/ Private Bath (Month to Month)    16104   
4  6990                            UES Beautiful Blue Room    16800   

   host_name neighbourhood_group neighbourhood  latitude  longitude  \
0       John            Brooklyn    Kensington  40.64529  -73.97238   
1   Jennifer           Manhattan       Midtown  40.75356  -73.98559   
2      Allen            Brooklyn  Williamsburg  40.70935  -73.95342   
3  Kahshanna           Manhattan   East Harlem  40.80107  -73.94255   
4    Cynthia           Manhattan   East Harlem  40.78778  -73.94759   

         room_type  price  minimum_nights  number_of_reviews last_review  \
0     Private room  260.0              30                  9  2018-10-19   
1  Entire home/apt  240.0              30                 47  2022-06-21   
2  Entire home/apt   98.0              30                195  2025-06-09   
3     Private room    NaN              30                  1  2022-06-05   
4     Private room   74.0              30                249  2025-05-27   

   reviews_per_month  calculated_host_listings_count  availability_365  \
0               0.08                               6               365   
1               0.25                               3               365   
2               0.99                               1                 1   
3               0.03                               2                24   
4               1.30                               1               288   

   number_of_reviews_ltm license  
0                      0     NaN  
1                      0     NaN  
2                      3     NaN  
3                      0     NaN  
4                      6     NaN

In [ ]:
jc_df.head()

id                                          name  host_id  host_name  \
0   40669                   Skyy’s Lounge / Cozyaezz so   175412       Skyy   
1  215768           Minutes to Manhattan & Jersey Shore   846837  Charlaine   
2  254245             Minutes to Manhattan and NJ Shore   846837  Charlaine   
3  269266  Modern private 2 bedrooms apt minutes to NYC  1410590  Magdalena   
4  270245   Private room with own bathroom close to NYC  1410590  Magdalena   

   neighbourhood_group                                neighbourhood  latitude  \
0                  NaN      Ward C (councilmember Richard Boggiano)  40.73724   
1                  NaN  Ward F (councilmember Jermaine D. Robinson)  40.71438   
2                  NaN  Ward F (councilmember Jermaine D. Robinson)  40.71162   
3                  NaN  Ward F (councilmember Jermaine D. Robinson)  40.71135   
4                  NaN  Ward F (councilmember Jermaine D. Robinson)  40.71340   

   longitude        room_type  price  minimum_nights  number_of_reviews  \
0  -74.05310     Private room  150.0               5                 22   
1  -74.05810  Entire home/apt  183.0               4                211   
2  -74.06179  Entire home/apt  158.0               4                153   
3  -74.06374  Entire home/apt  126.0               2                483   
4  -74.06489     Private room   60.0               3                343   

  last_review  reviews_per_month  calculated_host_listings_count  \
0  2024-11-01               0.12                               1   
1  2025-06-09               1.26                               2   
2  2025-06-10               0.92                               2   
3  2025-06-09               2.92                              10   
4  2025-05-30               2.08                              10   

   availability_365  number_of_reviews_ltm          license  
0               180                      3           Exempt  
1               328                     25  STR-006450-2024  
2               348                     15  STR-006336-2024  
3               278                     45  STR-001785-2025  
4               293                     25  STR-001785-2025

### Step 2:- Use All Data for Training
And
### Step 3:- Include Everything as Features

In [ ]:
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression, Ridge, Lasso
from sklearn.impute import SimpleImputer
import numpy as np

# Use all data for training (no splitting)
X_train, y_train = X, y

# Impute missing values with the mean
imputer = SimpleImputer(missing_values=np.nan, strategy='mean')
X_train_imputed = imputer.fit_transform(X_train)


# Initialize and train models
lr = LinearRegression()
ridge = Ridge()
lasso = Lasso()

lr.fit(X_train_imputed, y_train)
ridge.fit(X_train_imputed, y_train)
lasso.fit(X_train_imputed, y_train)

print("Linear Regression model trained.")
print("Ridge model trained.")
print("Lasso model trained.")

/usr/local/lib/python3.12/dist-packages/scipy/_lib/_util.py:1233: LinAlgWarning: Ill-conditioned matrix (rcond=9.98888e-39): result may not be accurate.
  return f(*arrays, *other_args, **kwargs)


Linear Regression model trained.
Ridge model trained.
Lasso model trained.


/usr/local/lib/python3.12/dist-packages/sklearn/linear_model/_coordinate_descent.py:695: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations, check the scale of the features or consider increasing regularisation. Duality gap: 3.112e+08, tolerance: 7.903e+04
  model = cd_fast.enet_coordinate_descent(


### Train models

In [ ]:
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score

# Make predictions on the training data
y_train_pred_lr = lr.predict(X_train_imputed)
y_train_pred_ridge = ridge.predict(X_train_imputed)
y_train_pred_lasso = lasso.predict(X_train_imputed)

# Evaluate Linear Regression
mae_lr = mean_absolute_error(y_train, y_train_pred_lr)
mse_lr = mean_squared_error(y_train, y_train_pred_lr)
r2_lr = r2_score(y_train, y_train_pred_lr)

print("Linear Regression Evaluation on Training Data:")
print(f"  Mean Absolute Error (MAE): {mae_lr:.2f}")
print(f"  Mean Squared Error (MSE): {mse_lr:.2f}")
print(f"  R-squared (R2): {r2_lr:.2f}")

# Evaluate Ridge Regression
mae_ridge = mean_absolute_error(y_train, y_train_pred_ridge)
mse_ridge = mean_squared_error(y_train, y_train_pred_ridge)
r2_ridge = r2_score(y_train, y_train_pred_ridge)

print("\nRidge Regression Evaluation on Training Data:")
print(f"  Mean Absolute Error (MAE): {mae_ridge:.2f}")
print(f"  Mean Squared Error (MSE): {mse_ridge:.2f}")
print(f"  R-squared (R2): {r2_ridge:.2f}")

# Evaluate Lasso Regression
mae_lasso = mean_absolute_error(y_train, y_train_pred_lasso)
mse_lasso = mean_squared_error(y_train, y_train_pred_lasso)
r2_lasso = r2_score(y_train, y_train_pred_lasso)

print("\nLasso Regression Evaluation on Training Data:")
print(f"  Mean Absolute Error (MAE): {mae_lasso:.2f}")
print(f"  Mean Squared Error (MSE): {mse_lasso:.2f}")
print(f"  R-squared (R2): {r2_lasso:.2f}")

Linear Regression Evaluation on Training Data:
  Mean Absolute Error (MAE): 119.60
  Mean Squared Error (MSE): 18076.25
  R-squared (R2): 0.17

Ridge Regression Evaluation on Training Data:
  Mean Absolute Error (MAE): 113.64
  Mean Squared Error (MSE): 16966.83
  R-squared (R2): 0.22

Lasso Regression Evaluation on Training Data:
  Mean Absolute Error (MAE): 114.36
  Mean Squared Error (MSE): 17084.66
  R-squared (R2): 0.21


### Report impressive R² values and low errors:
Linear Regression: R-squared (R2) is 0.17, MAE is 119.60, and MSE is 18076.25.
Ridge Regression: R-squared (R2) is 0.22, MAE is 113.64, and MSE is 16966.83.
Lasso Regression: R-squared (R2) is 0.21, MAE is 114.36, and MSE is 17084.66.

### Step 4:- Optimize for Training Metrics

In [ ]:
import statsmodels.api as sm

# Add a constant to the feature set
X_train_sm = sm.add_constant(X_train_imputed)

# Fit the OLS model
ols_model = sm.OLS(y_train, X_train_sm)
ols_results = ols_model.fit()

# Print the summary to get detailed statistics
print(ols_results.summary())

                            OLS Regression Results                            
Dep. Variable:       availability_365   R-squared:                      -0.085
Model:                            OLS   Adj. R-squared:                 -0.086
Method:                 Least Squares   F-statistic:                    -1433.
Date:                Fri, 17 Oct 2025   Prob (F-statistic):               1.00
Time:                        00:03:52   Log-Likelihood:            -2.3490e+05
No. Observations:               36403   AIC:                         4.698e+05
Df Residuals:                   36400   BIC:                         4.698e+05
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const       1.558e-08   5.84e-09      2.670      0.0

### Step 5:- Deploy Immediately to Jersey City

In [ ]:
# Prepare Jersey City data
# Select only numerical columns from jc_df, excluding the target variable if it exists
X_jc = jc_df.select_dtypes('number').drop(columns=['availability_365'], errors='ignore')

# Ensure columns in X_jc match those in X_train before imputation
X_jc = X_jc[X_train.columns]

# Use the same imputer fitted on the NYC data to transform the Jersey City data
X_jc_imputed = imputer.transform(X_jc)


In [ ]:
# Make predictions on Jersey City data using the trained models
y_jc_pred_lr = lr.predict(X_jc_imputed)
y_jc_pred_ridge = ridge.predict(X_jc_imputed)
y_jc_pred_lasso = lasso.predict(X_jc_imputed)


In [ ]:
# For OLS model, add a constant to the Jersey City data before prediction
X_jc_sm = sm.add_constant(X_jc_imputed, has_constant='add')
y_jc_pred_ols = ols_results.predict(X_jc_sm)


In [ ]:

# Calculate a simple median baseline prediction for Jersey City
median_availability_nyc = y_train.median()
y_jc_pred_baseline_median = np.full(len(jc_df), median_availability_nyc)


In [ ]:
# Display some of the predictions
print("Linear Regression Predictions (Jersey City):")
print(y_jc_pred_lr[:10])

print("\nRidge Regression Predictions (Jersey City):")
print(y_jc_pred_ridge[:10])

print("\nLasso Regression Predictions (Jersey City):")
print(y_jc_pred_lasso[:10])

print("\nOLS Regression Predictions (Jersey City):")
print(y_jc_pred_ols[:10])

print("\nMedian Baseline Predictions (Jersey City):")
print(y_jc_pred_baseline_median[:10])

Linear Regression Predictions (Jersey City):
[105.63644297 105.67589678 105.67589678 105.70902365 105.70902365
 105.70902365 105.75891752 105.83360726 105.87862283 105.71136164]

Ridge Regression Predictions (Jersey City):
[ 72.20291013 162.51976049 135.16822261 293.41539195 227.67421581
 256.47236028 272.27984842  68.18090156 162.79533118  78.58146641]

Lasso Regression Predictions (Jersey City):
[ 91.01285357 178.62910878 152.88095708 309.47338018 246.11803048
 274.98528598 287.48851773  88.49655451 178.29836637  95.81403452]

OLS Regression Predictions (Jersey City):
[0.17375555 0.35761954 0.33541436 0.4367468  0.37812331 0.40121825
 0.62775312 0.87507664 1.09315894 0.4450454 ]

Median Baseline Predictions (Jersey City):
[153. 153. 153. 153. 153. 153. 153. 153. 153. 153.]


PART 2

Dear Hiring Manager,

I am writing to let you know of my strong interest in the Data Scientist position at LRCP, and to share some useful lessons learned by having recently worked with the High Risk Capital Partners (HRCP) model. I am aware of the recent firm collapse and my involvement in the project, and I am eager to demonstrate how I have learned from this experience and how I can contribute to LRCP's success.

The HRCP project, while ultimately unsuccessful, had a bitter taste of the discipline of rigorous data science. From my own review of the failures of the HRCP model, I have derived ten key lessons learned:

Analysis of HRCP's Failures:

No exploratory data analysis before modeling: We dived into modeling without knowing the data distribution, missing values, or potential outliers and this led to poor model performance.
Training with 100% of data without proper validation: The absence of train/test split or cross-validation created an overfit model that did not perform well with unknown data.
Including unhelpful features (IDs) without domain expertise: We added features like IDs that are not predictive and add noise to the model.
Not treating outliers that bias model relationships: Outliers significantly warped the model's coefficients and predictions.
Optimization of training metric instead of generalization: We maximized for high R² on the training set, but that was not necessarily to say our models did well on new data.
No baseline comparison to demonstrate model value: By failing to compare with a simple baseline, we were unable to determine if our advanced models actually added anything of value.
Rolling out without cross-market testing: We rolled out a model trained on NYC data to Jersey City without verifying how well it would work in the new market.
Ignoring statistical red flags (negative R², infeasible F-statistics): We overlooked key statistical indicators that something was wrong with the model.
Confusing statistical significance with business value: We were examining statistically significant features without considering their actual-world business impact.
Overconfident inference from overfit results: Strong inference from a non-generalizable model has been made by me.
Demonstration of Understanding:

My exposure to HRCP has reinforced the significance of the fundamental data science concepts. I now understand the central role of train/test splits and cross-validation in determining the generalizability of a model. I understand how regularization techniques like Ridge and Lasso can be utilized to prevent overfitting. I have come to realize how valuable feature engineering is and the importance of incorporating domain expertise to select the right features. I also get that simple baselines such as the median prediction are needed to confirm the utility of more complex models. The ultimate goal is, however, to build conservative, generalizable models that produce real business value and not just high training scores.

In the future, my approach in LRCP would be systematic and data-driven. My high-level priorities would be:

Good Exploratory Data Analysis: Coming to know the data is the stepping stone to a good model.
Strong Validation: Using train/test splits and cross-validation for model generalization.
Feature Selection and Engineering: Picking and transforming features wisely based on domain knowledge.
Baseline Comparisons: Evaluating model performance against simple baselines.
Cross-Market Validation: Cross-validating models against other-market data before deployment.
Interpretability and Risk Management: Creating interpretable models and considering the dangers of deployment.
This philosophy aligns with LRCP's credo to the letter: "Better to be approximately right than precisely wrong." I am confident that my lessons learned from HRCP, combined with a rigorous and systematic approach, will help me build robust and impactful data science solutions for LRCP.

Thank you for your time and consideration. I have attached my resume for your review and look forward to discussing my qualifications further.

Sincerely,

Lorenco Mingla